# Data Exploration and Cleaning
## US Clinical Provider 2009

* Source: National Plan and Provider Enumeration System (NPPES) | NBER
* Data sets: 
    - Country.csv
    - State.csv
    - Specialty.csv
    - Taxonomy.csv
    - Provider.csv
    - Followup2009-1.csv
    - Followup2009-2.csv
    - Followup2009-3.csv
    - Followup2009-4.csv
    - Followup2009-5.csv
    - Followup2009-6.csv
    - Followup2009-7.csv

In [1]:
# Import libraries
import pandas as pd
from csv import QUOTE_ALL

Since we havev many files, and codes is repeatedly called, we have created a function for common data exploration.

In [3]:
# Function to check null vavlues
def check_null(data):
    null_check = data.isnull().sum()
    print(f'\nNull values:\n{null_check}')
    for c,v in null_check.items():
        if v > 0:
            print(v)
            display(data[(data[c].isnull())])

# Function for exploration
def explore_data(file, data=None, subset=None):
    if file:
        data = pd.read_csv(file)
    else:
        if data is None:
            print('Please pass file or data frame.')
            return None
    print(f'\n\n{file} Exploration:')
    
    # Check data types
    print('\nData types:')
    print(data.info())

    # Null check
    check_null(data)

    # Duplicate check
    print('\nDuplicate values:\n', data[(data.duplicated(subset=subset))])
    print('\nData overview:')
    display(data.head())
    return data

### Country.csv
* Check data types
* Check null values
* Check duplicates

In [4]:
country = explore_data('data\Extract2009-FullDoubleQuotes\Country.csv')



..\..\data\Extract2009-FullDoubleQuotes\Country.csv Exploration:

Data types:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 235 entries, 0 to 234
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   Code    234 non-null    object
 1   Name    235 non-null    object
dtypes: object(2)
memory usage: 3.8+ KB
None

Null values:
Code    1
Name    0
dtype: int64
1


Code     Name
151  NaN  NAMIBIA


Duplicate values:
 Empty DataFrame
Columns: [Code, Name]
Index: []

Data overview:


Code                  Name
0   AD               ANDORRA
1   AE  UNITED ARAB EMIRATES
2   AF           AFGHANISTAN
3   AG   ANTIGUA AND BARBUDA
4   AI              ANGUILLA

From the report above, there is null values of NAMIBIA country code and we can drop it since analysis is for US only.

In [5]:
country.dropna(inplace=True)
check_null(country)


Null values:
Code    0
Name    0
dtype: int64


Saved clean data to file {file}_clean.csv

In [6]:
country.to_csv(
    'Country_clean.csv',
    index=False,
    quoting=QUOTE_ALL
)

### State.csv
* Check data types
* Check null values
* Check duplicates

In [7]:
state = explore_data('data\Extract2009-FullDoubleQuotes\State.csv')



..\..\data\Extract2009-FullDoubleQuotes\State.csv Exploration:

Data types:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 60 entries, 0 to 59
Data columns (total 3 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   Code        60 non-null     object
 1   Name        60 non-null     object
 2   Governance  60 non-null     object
dtypes: object(3)
memory usage: 1.5+ KB
None

Null values:
Code          0
Name          0
Governance    0
dtype: int64

Duplicate values:
 Empty DataFrame
Columns: [Code, Name, Governance]
Index: []

Data overview:


Code            Name Governance
0   AK          ALASKA      STATE
1   AL         ALABAMA      STATE
2   AR        ARKANSAS      STATE
3   AS  AMERICAN SAMOA  TERRITORY
4   AZ         ARIZONA      STATE

In State file, no duplicates and null values to clean up.

### Specialty.csv
* Check data types
* Check null values
* Check duplicates

In [8]:
specialty = explore_data('data\Extract2009-FullDoubleQuotes\Specialty.csv')



..\..\data\Extract2009-FullDoubleQuotes\Specialty.csv Exploration:

Data types:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7502413 entries, 0 to 7502412
Data columns (total 2 columns):
 #   Column    Dtype 
---  ------    ----- 
 0   Provider  int64 
 1   Code      object
dtypes: int64(1), object(1)
memory usage: 114.5+ MB
None

Null values:
Provider    0
Code        0
dtype: int64

Duplicate values:
 Empty DataFrame
Columns: [Provider, Code]
Index: []

Data overview:


Provider        Code
0  1003000100  171M00000X
1  1003000118  208000000X
2  1003000126  207R00000X
3  1003000134  207ZP0102X
4  1003000142  207L00000X

In Specialty file, no duplicates and null values to clean up.

### Taxonomy.csv
* Check data types
* Check null values
* Check duplicates

In [8]:
taxo = explore_data('data\Extract2009-FullDoubleQuotes\Taxonomy.csv')



..\..\data\Extract2009-FullDoubleQuotes\Taxonomy.csv Exploration:

Data types:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 874 entries, 0 to 873
Data columns (total 6 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   Code            874 non-null    object
 1   Grouping        874 non-null    object
 2   Classification  874 non-null    object
 3   Specialization  633 non-null    object
 4   Name            874 non-null    object
 5   Individual      874 non-null    bool  
dtypes: bool(1), object(5)
memory usage: 35.1+ KB
None

Null values:
Code                0
Grouping            0
Classification      0
Specialization    241
Name                0
Individual          0
dtype: int64
241


Code                             Grouping  \
0    193200000X                                Group   
1    193400000X                                Group   
2    207K00000X  Allopathic & Osteopathic Physicians   
5    207L00000X  Allopathic & Osteopathic Physicians   
11   208U00000X  Allopathic & Osteopathic Physicians   
..          ...                                  ...   
869  343800000X              Transportation Services   
870  344600000X              Transportation Services   
871  347D00000X              Transportation Services   
872  347E00000X              Transportation Services   
873  342000000X              Transportation Services   

                      Classification Specialization  \
0                    Multi-Specialty            NaN   
1                   Single Specialty            NaN   
2               Allergy & Immunology            NaN   
5                     Anesthesiology            NaN   
11             Clinical Pharmacology            NaN   
..                               ...            ...   
869  Secured Medical Transport (VAN)            NaN   
870                             Taxi            NaN   
871                            Train            NaN   
872            Transportation Broker            NaN   
873   Transportation Network Company            NaN   

                                Name  Individual  
0              Multi-Specialty Group        True  
1             Single Specialty Group        True  
2     Allergy & Immunology Physician        True  
5           Anesthesiology Physician        True  
11   Clinical Pharmacology Physician        True  
..                               ...         ...  
869  Secured Medical Transport (VAN)       False  
870                             Taxi       False  
871                            Train       False  
872            Transportation Broker       False  
873   Transportation Network Company       False  

[241 rows x 6 columns]


Duplicate values:
 Empty DataFrame
Columns: [Code, Grouping, Classification, Specialization, Name, Individual]
Index: []

Data overview:


Code                             Grouping        Classification  \
0  193200000X                                Group       Multi-Specialty   
1  193400000X                                Group      Single Specialty   
2  207K00000X  Allopathic & Osteopathic Physicians  Allergy & Immunology   
3  207KA0200X  Allopathic & Osteopathic Physicians  Allergy & Immunology   
4  207KI0005X  Allopathic & Osteopathic Physicians  Allergy & Immunology   

                     Specialization  \
0                               NaN   
1                               NaN   
2                               NaN   
3                           Allergy   
4  Clinical & Laboratory Immunology   

                                                Name  Individual  
0                              Multi-Specialty Group        True  
1                             Single Specialty Group        True  
2                     Allergy & Immunology Physician        True  
3                                  Allergy Physician        True  
4  Clinical & Laboratory Immunology (Allergy & Im...        True

In Taxonomy file, we can see that there are 241 null values in Specialization column, we will not need to clean this up. The field that we will reference into this file is the Code column.

### Provider.csv
* Check data types
* Check null values
* Check duplicates

In [13]:
provider = explore_data('data\Extract2009-FullDoubleQuotes\Provider.csv')

data = pd.read_csv('data\Extract2009-FullDoubleQuotes\Provider.csv')
data = data[(data['Country']=='US')]
explore_data(file=None, data=data)

nulldatas = data[(data['City'].isnull())]
nulldatas['Individual'].value_counts()



..\..\data\Extract2009-FullDoubleQuotes\Provider.csv Exploration:

Data types:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7794261 entries, 0 to 7794260
Data columns (total 6 columns):
 #   Column      Dtype 
---  ------      ----- 
 0   Provider    int64 
 1   Individual  bool  
 2   Zip         object
 3   City        object
 4   State       object
 5   Country     object
dtypes: bool(1), int64(1), object(4)
memory usage: 304.8+ MB
None

Null values:
Provider           0
Individual         0
Zip           247647
City          247643
State         247671
Country       247643
dtype: int64
247647


Provider  Individual  Zip City State Country
121      1003001314       False  NaN  NaN   NaN     NaN
150      1003001603       False  NaN  NaN   NaN     NaN
151      1003001611       False  NaN  NaN   NaN     NaN
156      1003001660       False  NaN  NaN   NaN     NaN
169      1003001793       False  NaN  NaN   NaN     NaN
...             ...         ...  ...  ...   ...     ...
7794160  1992998892       False  NaN  NaN   NaN     NaN
7794177  1992999064       False  NaN  NaN   NaN     NaN
7794205  1992999346       False  NaN  NaN   NaN     NaN
7794216  1992999452       False  NaN  NaN   NaN     NaN
7794242  1992999718       False  NaN  NaN   NaN     NaN

[247647 rows x 6 columns]

247643


Provider  Individual  Zip City State Country
121      1003001314       False  NaN  NaN   NaN     NaN
150      1003001603       False  NaN  NaN   NaN     NaN
151      1003001611       False  NaN  NaN   NaN     NaN
156      1003001660       False  NaN  NaN   NaN     NaN
169      1003001793       False  NaN  NaN   NaN     NaN
...             ...         ...  ...  ...   ...     ...
7794160  1992998892       False  NaN  NaN   NaN     NaN
7794177  1992999064       False  NaN  NaN   NaN     NaN
7794205  1992999346       False  NaN  NaN   NaN     NaN
7794216  1992999452       False  NaN  NaN   NaN     NaN
7794242  1992999718       False  NaN  NaN   NaN     NaN

[247643 rows x 6 columns]

247671


Provider  Individual  Zip City State Country
121      1003001314       False  NaN  NaN   NaN     NaN
150      1003001603       False  NaN  NaN   NaN     NaN
151      1003001611       False  NaN  NaN   NaN     NaN
156      1003001660       False  NaN  NaN   NaN     NaN
169      1003001793       False  NaN  NaN   NaN     NaN
...             ...         ...  ...  ...   ...     ...
7794160  1992998892       False  NaN  NaN   NaN     NaN
7794177  1992999064       False  NaN  NaN   NaN     NaN
7794205  1992999346       False  NaN  NaN   NaN     NaN
7794216  1992999452       False  NaN  NaN   NaN     NaN
7794242  1992999718       False  NaN  NaN   NaN     NaN

[247671 rows x 6 columns]

247643


Provider  Individual  Zip City State Country
121      1003001314       False  NaN  NaN   NaN     NaN
150      1003001603       False  NaN  NaN   NaN     NaN
151      1003001611       False  NaN  NaN   NaN     NaN
156      1003001660       False  NaN  NaN   NaN     NaN
169      1003001793       False  NaN  NaN   NaN     NaN
...             ...         ...  ...  ...   ...     ...
7794160  1992998892       False  NaN  NaN   NaN     NaN
7794177  1992999064       False  NaN  NaN   NaN     NaN
7794205  1992999346       False  NaN  NaN   NaN     NaN
7794216  1992999452       False  NaN  NaN   NaN     NaN
7794242  1992999718       False  NaN  NaN   NaN     NaN

[247643 rows x 6 columns]


Duplicate values:
 Empty DataFrame
Columns: [Provider, Individual, Zip, City, State, Country]
Index: []

Data overview:


Provider  Individual        Zip         City State Country
0  1003000100        True  900053200  LOS ANGELES    CA      US
1  1003000118       False  285602304     NEW BERN    NC      US
2  1003000126        True  208171841     BETHESDA    MD      US
3  1003000134        True  602011718     EVANSTON    IL      US
4  1003000142        True  436233536       TOLEDO    OH      US



None Exploration:

Data types:
<class 'pandas.core.frame.DataFrame'>
Int64Index: 7541570 entries, 0 to 7794260
Data columns (total 6 columns):
 #   Column      Dtype 
---  ------      ----- 
 0   Provider    int64 
 1   Individual  bool  
 2   Zip         object
 3   City        object
 4   State       object
 5   Country     object
dtypes: bool(1), int64(1), object(4)
memory usage: 352.4+ MB
None

Null values:
Provider      0
Individual    0
Zip           4
City          0
State         2
Country       0
dtype: int64
4


Provider  Individual  Zip          City State Country
71555    1003935214       False  NaN          NAPA    CA      US
2409788  1306933650       False  NaN   WALLA WALLA    WA      US
4156118  1538348511       False  NaN     SAN DIMAS    CA      US
5464841  1700097797       False  NaN  PHILADELPHIA    PA      US

2


Provider  Individual        Zip      City State Country
4634116  1598356685        True  100250000  NEW YORK   NaN      US
5323805  1689229775        True  681040000     OMAHA   NaN      US


Duplicate values:
 Empty DataFrame
Columns: [Provider, Individual, Zip, City, State, Country]
Index: []

Data overview:


Provider  Individual        Zip         City State Country
0  1003000100        True  900053200  LOS ANGELES    CA      US
1  1003000118       False  285602304     NEW BERN    NC      US
2  1003000126        True  208171841     BETHESDA    MD      US
3  1003000134        True  602011718     EVANSTON    IL      US
4  1003000142        True  436233536       TOLEDO    OH      US

Series([], Name: Individual, dtype: int64)

In Provider file, we can see that there is no duplicate values but has many null values in Zip, City, State and Country. We can ignore this as SME informed that there are Providers that has no address details due to privacy.

### Followup2009-1.csv to Followup2009-7.csv
* Check data types
* Check null values
* Check duplicates

In [10]:
fu1_df = explore_data('data\Extract2009-FullDoubleQuotes\Followup2009-1.csv')



..\..\data\Extract2009-FullDoubleQuotes\Followup2009-1.csv Exploration:

Data types:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7500000 entries, 0 to 7499999
Data columns (total 5 columns):
 #   Column                  Dtype
---  ------                  -----
 0   Source Provider         int64
 1   Target Provider         int64
 2   Followups Within Month  int64
 3   Followups Within Day    int64
 4   Unique Clients          int64
dtypes: int64(5)
memory usage: 286.1 MB
None

Null values:
Source Provider           0
Target Provider           0
Followups Within Month    0
Followups Within Day      0
Unique Clients            0
dtype: int64

Duplicate values:
 Empty DataFrame
Columns: [Source Provider, Target Provider, Followups Within Month, Followups Within Day, Unique Clients]
Index: []

Data overview:


Source Provider  Target Provider  Followups Within Month  \
0        135632034       1356323034                      36   
1        135632034       1508848029                      96   
2        135632034       1700871498                      54   
3        135632034       1992786743                      48   
4       1000000004       1225194574                      55   

   Followups Within Day  Unique Clients  
0                     0              18  
1                    28              26  
2                     2              19  
3                     2              17  
4                    11              16

In [11]:
fu2_df = explore_data('data\Extract2009-FullDoubleQuotes\Followup2009-2.csv')



..\..\data\Extract2009-FullDoubleQuotes\Followup2009-2.csv Exploration:

Data types:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7500000 entries, 0 to 7499999
Data columns (total 5 columns):
 #   Column                  Dtype
---  ------                  -----
 0   Source Provider         int64
 1   Target Provider         int64
 2   Followups Within Month  int64
 3   Followups Within Day    int64
 4   Unique Clients          int64
dtypes: int64(5)
memory usage: 286.1 MB
None

Null values:
Source Provider           0
Target Provider           0
Followups Within Month    0
Followups Within Day      0
Unique Clients            0
dtype: int64

Duplicate values:
 Empty DataFrame
Columns: [Source Provider, Target Provider, Followups Within Month, Followups Within Day, Unique Clients]
Index: []

Data overview:


Source Provider  Target Provider  Followups Within Month  \
0       1134324262       1972681542                      27   
1       1134324262       1972684249                      34   
2       1134324338       1154314789                      34   
3       1134324494       1033174123                      15   
4       1134324494       1063497451                      28   

   Followups Within Day  Unique Clients  
0                     8              14  
1                    17              16  
2                    13              12  
3                     0              11  
4                     0              18

In [12]:
fu3_df = explore_data('data\Extract2009-FullDoubleQuotes\Followup2009-3.csv')



..\..\data\Extract2009-FullDoubleQuotes\Followup2009-3.csv Exploration:

Data types:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7500000 entries, 0 to 7499999
Data columns (total 5 columns):
 #   Column                  Dtype
---  ------                  -----
 0   Source Provider         int64
 1   Target Provider         int64
 2   Followups Within Month  int64
 3   Followups Within Day    int64
 4   Unique Clients          int64
dtypes: int64(5)
memory usage: 286.1 MB
None

Null values:
Source Provider           0
Target Provider           0
Followups Within Month    0
Followups Within Day      0
Unique Clients            0
dtype: int64

Duplicate values:
 Empty DataFrame
Columns: [Source Provider, Target Provider, Followups Within Month, Followups Within Day, Unique Clients]
Index: []

Data overview:


Source Provider  Target Provider  Followups Within Month  \
0       1285618991       1336145804                      45   
1       1285618991       1336149137                      89   
2       1285618991       1336172907                     413   
3       1285618991       1336184993                      39   
4       1285618991       1336251784                      55   

   Followups Within Day  Unique Clients  
0                     1              20  
1                     4              21  
2                   157             114  
3                     0              29  
4                     0              15

In [13]:
fu4_df = explore_data('data\Extract2009-FullDoubleQuotes\Followup2009-4.csv')



..\..\data\Extract2009-FullDoubleQuotes\Followup2009-4.csv Exploration:

Data types:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7500000 entries, 0 to 7499999
Data columns (total 5 columns):
 #   Column                  Dtype
---  ------                  -----
 0   Source Provider         int64
 1   Target Provider         int64
 2   Followups Within Month  int64
 3   Followups Within Day    int64
 4   Unique Clients          int64
dtypes: int64(5)
memory usage: 286.1 MB
None

Null values:
Source Provider           0
Target Provider           0
Followups Within Month    0
Followups Within Day      0
Unique Clients            0
dtype: int64

Duplicate values:
 Empty DataFrame
Columns: [Source Provider, Target Provider, Followups Within Month, Followups Within Day, Unique Clients]
Index: []

Data overview:


Source Provider  Target Provider  Followups Within Month  \
0       1427118595       1881817237                      21   
1       1427118595       1891794210                     313   
2       1427118595       1891816245                      19   
3       1427118595       1891866984                      55   
4       1427118595       1891946117                      45   

   Followups Within Day  Unique Clients  
0                     1              14  
1                    22              66  
2                     5              12  
3                     2              47  
4                     8              27

In [14]:
fu5_df = explore_data('data\Extract2009-FullDoubleQuotes\Followup2009-5.csv')



..\..\data\Extract2009-FullDoubleQuotes\Followup2009-5.csv Exploration:

Data types:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7500000 entries, 0 to 7499999
Data columns (total 5 columns):
 #   Column                  Dtype
---  ------                  -----
 0   Source Provider         int64
 1   Target Provider         int64
 2   Followups Within Month  int64
 3   Followups Within Day    int64
 4   Unique Clients          int64
dtypes: int64(5)
memory usage: 286.1 MB
None

Null values:
Source Provider           0
Target Provider           0
Followups Within Month    0
Followups Within Day      0
Unique Clients            0
dtype: int64

Duplicate values:
 Empty DataFrame
Columns: [Source Provider, Target Provider, Followups Within Month, Followups Within Day, Unique Clients]
Index: []

Data overview:


Source Provider  Target Provider  Followups Within Month  \
0       1578620449       1356326995                    1079   
1       1578620449       1356328991                     918   
2       1578620449       1356331276                      36   
3       1578620449       1356331326                     569   
4       1578620449       1356332431                      91   

   Followups Within Day  Unique Clients  
0                     0             227  
1                     0             436  
2                     0              17  
3                     0             289  
4                     0              53

In [15]:
fu6_df = explore_data('data\Extract2009-FullDoubleQuotes\Followup2009-6.csv')



..\..\data\Extract2009-FullDoubleQuotes\Followup2009-6.csv Exploration:

Data types:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7500000 entries, 0 to 7499999
Data columns (total 5 columns):
 #   Column                  Dtype
---  ------                  -----
 0   Source Provider         int64
 1   Target Provider         int64
 2   Followups Within Month  int64
 3   Followups Within Day    int64
 4   Unique Clients          int64
dtypes: int64(5)
memory usage: 286.1 MB
None

Null values:
Source Provider           0
Target Provider           0
Followups Within Month    0
Followups Within Day      0
Unique Clients            0
dtype: int64

Duplicate values:
 Empty DataFrame
Columns: [Source Provider, Target Provider, Followups Within Month, Followups Within Day, Unique Clients]
Index: []

Data overview:


Source Provider  Target Provider  Followups Within Month  \
0       1730105719       1093720385                     113   
1       1730105719       1093728388                     196   
2       1730105719       1093733412                     594   
3       1730105719       1093735763                     674   
4       1730105719       1093742249                      27   

   Followups Within Day  Unique Clients  
0                     0              62  
1                     0              90  
2                     0             292  
3                     0             316  
4                     0              18

In [16]:
fu7_df = explore_data('data\Extract2009-FullDoubleQuotes\Followup2009-7.csv')



..\..\data\Extract2009-FullDoubleQuotes\Followup2009-7.csv Exploration:

Data types:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5382951 entries, 0 to 5382950
Data columns (total 5 columns):
 #   Column                  Dtype
---  ------                  -----
 0   Source Provider         int64
 1   Target Provider         int64
 2   Followups Within Month  int64
 3   Followups Within Day    int64
 4   Unique Clients          int64
dtypes: int64(5)
memory usage: 205.3 MB
None

Null values:
Source Provider           0
Target Provider           0
Followups Within Month    0
Followups Within Day      0
Unique Clients            0
dtype: int64

Duplicate values:
 Empty DataFrame
Columns: [Source Provider, Target Provider, Followups Within Month, Followups Within Day, Unique Clients]
Index: []

Data overview:


Source Provider  Target Provider  Followups Within Month  \
0       1881704807       1154304053                      67   
1       1881704807       1154391456                      91   
2       1881704807       1164412276                      28   
3       1881704807       1164607669                      46   
4       1881704807       1174508717                      81   

   Followups Within Day  Unique Clients  
0                     0              43  
1                     0              41  
2                     0              24  
3                     0              37  
4                     0              15

In Followup2009-1 to Followup2009-7 files, no duplicate and null values.